solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -148,29 +127,186 @@ library Groups {
+                // it from the activeTerminatedGroups array manually by rearranging
+                // the order starting from 'i'th group.
+                for (
+                    uint256 j = i;
+                    j < self.activeTerminatedGroups.length - 1;
+                    j++
+                ) {
+                    self.activeTerminatedGroups[j] = self
+                        .activeTerminatedGroups[j + 1];
+                }
+                // Resizing the array length by 1. The last element was copied
+                // over in the loop above to an index "second to last". This is
+                // why we can safely remove it from here.
+                self.activeTerminatedGroups.pop();
+                // At this point the length of activeTerminatedGroups[] was shrinked
+                // by 1. We need to adjust 'i'th counter by 1 as well, otherwise
+                // it will overflow this array and throw an error.
+                i--;
+            }
+        }
+    }
+
+    /// @notice Terminates group with the provided index. Reverts if the group
+    ///         is already terminated.
+    /// @param  groupId Index in the groupRegistry array.
+    function terminateGroup(Data storage self, uint64 groupId) internal {
+        require(
+            !isGroupTerminated(self, groupId),
+            "Group has been already terminated"
+        );
+        self.groupsData[self.groupsRegistry[groupId]].terminated = true;
+        self.activeTerminatedGroups.push();
+        self.activeGroupsCount--;
+
+        // Sorting activeTerminatedGroups in ascending order so a non-terminated
+        // group is properly selected.
+        uint256 i;
+        for (
+            i = self.activeTerminatedGroups.length - 1;
+            i > 0 && self.activeTerminatedGroups[i - 1] > groupId;
+            i--
+        ) {
+            self.activeTerminatedGroups[i] = self.activeTerminatedGroups[i - 1];
+        }
+        self.activeTerminatedGroups[i] = groupId;
+    }
+
+    /// @notice Returns an index of a randomly selected active group. Terminated
+    ///         and expired groups are not considered as active.
+    ///         Before new group is selected, information about expired groups
+    ///         is updated. At least one active group needs to be present for this
+    ///         function to succeed.
+    /// @param seed Random number used as a group selection seed.
+    function selectGroup(Data storage self, uint256 seed)
         internal
-        view
         returns (uint64)
     {
-        // TODO: Revisit and include candidate, terminated and expired groups
-        return self.activeGroupsCount;
-        // TODO: Subtract expired and terminated groups
-        // .sub(self.expiredGroupOffset).sub(
-        //     self.activeTerminatedGroups.length)
+        expireOldGroups(self);
+
+        require(numberOfActiveGroups(self) > 0, "No active groups");
```

Comment:
If I'm not wrong this `require` combined with the logic living in `RandomBeacon.reportRelayEntryTimeout` can block the beacon in certain circumstances. Suppose the following scenario:

1. We have two groups A and B
2. Group A is almost expired, group B handles a relay request
3. Group B fails to provide the entry and the hard timeout is hit
4. Someone wants to report the timeout and calls `RandomBeacon.reportRelayEntryTimeout`
5. During the call ^, group B gets terminated and `groups.numberOfActiveGroups() > 0` condition is true so we want to enter `relay.retryOnEntryTimeout` but before we need to select the new group 
6. Unfortunately, group A exceeded its lifetime so during the `groups.selectGroup` it gets expired during `expireOldGroups` call
7. We revert in `require(numberOfGroups(self) > 0, "No active groups");`

In that case, nobody will be able to report the entry timeout. Please double-check because maybe I miss something but if that scenario is possible, we need to avoid reverting during `selectGroup`.

Reply:
Ok, let's add some block numbers to your scenario to see it clearly. What we have here strictly for group selection is 1:1 from beacon v1. There might be some diff how we call it upstream, but from what I can see it is very similar to v1 as well.
- say the relay hard timeout is at block 10
- group B expires at block 15
- group A expires at block 10
4) now at block# 10 and someone called `RandomBeacon.reportRelayEntryTimeout`
5) B is terminated
6) group A (at block 10) is expired
7) revert
Is this correct?

If so, I guess we have at least 3 possible routes:
a) this is what we want and revert is valid, cause there are no active groups.
b) use `if` instead of `require`
c) count the number of actual active groups on the fly "ahead" of time and enter the `cleanupOnEntryTimeout` instead of `retryOnEntryTimeout`. The way it works now is that say all the groups technically expire at block 10. Now it is block # 100. And even though they are all expired "on paper", before we call the 'selectGroup' they are all marked as "active".

#pdyraga thoughts?

Reply:
Yep, that's correct. I see it's from v1 but in v1 we had a very similar problem that blocked the beacon. Regarding proposed solutions:
- Option "a" will block the beacon as we won't be able to report a timeout due to permanent reverts
- Option "b" also won't work. We could return `0` as result of `selectGroup` in case there are no groups but `0` is a valid `groupId`
- Option "c" seems the most promising. Is there an easy way to count the groups "ahead"?

@pdyraga Have you any other ideas?

Reply:
I think c) but isn't it all about calling `expireOldGroups` in `reportRelayEntryTimeout` after the group is terminated so that `numberOfGroups(self) == 0` one line later?

Reply:
Yep, that's an option but we probably need to decouple `expireOldGroups` from `selectGroup` as well. It depends on the gas cost since `submitRelayEntry` already exceeds our desired limit.

Reply:
I am still a bit out of context with this code so pardon my potential ignorance, but my understanding is that `selectGroup` in most cases happens in `requestRelayEntry`, so it is unrelated to `submitRelayEntry` gas cost. Another occurrence should be `reportRelayEntryTimeout` transaction that needs to potentially select another group. What did I miss?

Reply:
No, you are right. That's me who mixed up things, sorry 🤦

Reply:
I think what Piotr suggested is the easiest, yet slipped my radar. So `expireOldGroups()` right after the `terminateGroup()` in `reportRelayEntryTimeout`

```
 function reportRelayEntryTimeout() external {
        (..)

        groups.terminateGroup()
        groups.expireOldGroups()

        if (groups.numberOfActiveGroups() > 0) {
            (..)
        } else {
            (..)
        }
    }
```

And yes, we have 1 another place that calls `selectGroup()` in `requestRelayEntry()`. We can leave `selectGroup()` as is or decouple `expireOldGroups()` and move to `requestRelayEntry()`. I think both are fine.

Reply:
Up to you 👍

Reply:
Added `groups.expireOldGroups()` + tests f874263. The rest is untouched.